client/homebrew/editor/editor.jsx

```diff
@@ -174,9 +174,39 @@ const Editor = createClass({
 		}
 	},
 
-	brewJump : function(){
-		const currentPage = this.getCurrentPage();
-		window.location.hash = `p${currentPage}`;
+	brewJump : function(targetPage=this.getCurrentPage()){
+		if(!window || this.isMeta()) return;
+		window.frames['BrewRenderer'].contentDocument.getElementById(`p${targetPage}`).scrollIntoView({ behaviour: 'auto', block: 'start' });
+		// const hashPage = (page != 1) ? `p${page}` : '';
+		// window.location.hash = hashPage;
+	},
+
+	sourceJump : function(targetLine=null){
+		if(this.isText() || this.isStyle()) {
+			if(targetLine == null) {
+				const brewPosition = window.frames['BrewRenderer'].contentDocument.getElementsByClassName('brewRenderer').item(0).scrollTop;
+				const pageCollection = window.frames['BrewRenderer'].contentDocument.getElementsByClassName('page');
+
+				let currentPage = 0;
+				let currentPagePosition = 0;
+				while (currentPagePosition < brewPosition && currentPage < pageCollection.length) {
+					currentPagePosition = currentPagePosition + pageCollection[currentPage].scrollHeight;
+					currentPage++;
+				};
```

Comment:
We do a similar calculation for this in the `handleScroll()` function in `brewRenderer.jsx` but it's actually pretty simple (excerpted and modified for legibility):

```
const target = e.target;
const currentPage = Math.floor(target.scrollTop / target.scrollHeight * prevState.pages.length)
```

Where `e` is the `brewRenderer` DIV element.

Reply:
As I understand it, `Math.floor(target.scrollTop / target.scrollHeight * pages.length)` finds the fraction of the BrewRenderer height and multiplies it by the number of pages, truncating the result to zero decimal places. This approach then would be inaccurate for pages of differing sizes - for example, if some pages were in portrait and others in landscape.
This method should (bearing in mind @ericscheid's comment about using getBoundingRect() instead) more accurately track the top of each page.

But, with that said, how often do users change the page size halfway through the brew? The `Math.floor()` method has been accurate enough for a long time now, but these functions will shine a light on any inaccuracy.

Reply:
Good point about potentially different page sizes. I do like the `boundingRect` approach better.